### DISCUSSION TITLE
HTML Widget

### DISCUSSION THREAD

**Discussion by misterek** [Textualize/textual#3104] [General]
Does there happy to be an HTML widget of any form?  I found https://github.com/HakierGrzonzo/html2rich, but was curious if there was anything built in that might help there.

**Comment by davep** [comment]
There's nothing in Textual itself, and I don't believe I've seen any third party library to do that. I suspect it would make for a useful library.

**Reply by misterek** [reply]
Thanks!  I'll close this, but maybe poke around a bit more with html2rich.

**Reply by TomJGooding** [reply]
I've toyed with the idea of creating a Textual-based web browser, but the problem is extracting the content from web pages without all the clutter.

I think you'd want something akin to the [Firefox Reader View](https://support.mozilla.org/en-US/kb/firefox-reader-view-clutter-free-web-pages) for displaying html content in the terminal. Unfortunately the few Python ports of the [readability library](https://github.com/mozilla/readability) I've tried aren't really good enough.

**Reply by misterek** [reply]
Yeah. My particular use case is mostly text based and pretty basic. Fortunately I can mostly just have BeautifulSoup make it into text. It's ugly, but less ugly than a bunch of markup in it

**Reply by TomJGooding** [reply]
You're welcome to borrow from my quickly hacked together attempt if it works for your simple use case: https://github.com/TomJGooding/textual-html

My idea was to convert html to markdown and nicely display in a [Markdown](https://textual.textualize.io/widgets/markdown/) widget. ~~Unfortunately I quickly found issues with the [markdownify](https://pypi.org/project/markdownify/) package mostly recommended for this, and it doesn't seem to be actively maintained.~~

I'm not sure I'm up to the task of creating both a decent readability-like filter _and_ html parser!

[EDIT: Sorry, edited to clarify that my issue seems to be with the readability port rather than markdownify after some further testing]

**Reply by misterek** [reply]
Wow. You work fast! I"ll play around with that when I have some time.   Nice approach, html->markdown.

**Reply by misterek** [reply]
Just to bring this around, I didn't use your exact code, but I didn't know that markdownify was a thing. That worked 100% perfectly for my use case. Thanks for that pointer!
